botorch/acquisition/hentropy_search.py

```diff
@@ -0,0 +1,275 @@
+#!/usr/bin/env python3
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+#
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+
+r"""
+This file has an implementation of Batch H-Entropy Search (qHES) via
+one-shot optimization as introduced in [Neiswanger2022]_.
+
+The authors adopt a generalized definition of entropy from past work
+in Bayesian decision theory, which proposes a family of decision-theoretic
+ entropies parameterized by a problem-specific loss function and
+ action set. Therefore, the method allow the development of a common
+ acquisition optimization procedure, which applies generically
+to many members of this family (where each member is induced by a
+specific loss function and action set).
+
+.. [Neiswanger2022]
+    W. Neiswanger, L. Yu, S. Zhao, C. Meng, S. Ermon. Generalizing Bayesian
+    Optimization with Decision-theoretic Entropies. Appears in Proceedings
+    of the 36th Conference on Neural Information Processing Systems
+    (NeurIPS 2022)
+"""
+
+import copy
+from typing import Any, Dict, Optional, Type
+
+import torch
+import torch.nn as nn
+from botorch.acquisition import MCAcquisitionObjective
+from botorch.acquisition.acquisition import OneShotAcquisitionFunction
+from botorch.acquisition.monte_carlo import MCAcquisitionFunction
+from botorch.models.model import Model
+from botorch.sampling.base import MCSampler
+from botorch.sampling.normal import SobolQMCNormalSampler
+from botorch.utils.transforms import t_batch_mode_transform
+from torch import Tensor
+
+
+class qHEntropySearch(MCAcquisitionFunction, OneShotAcquisitionFunction):
+    def __init__(
+        self,
+        model: Model,
+        loss_function_class: Type[MCAcquisitionObjective],
+        loss_function_hyperparameters: Dict[str, Any],
+        n_fantasy_at_design_pts: Optional[int] = 64,
+        n_fantasy_at_action_pts: Optional[int] = 64,
+        design_sampler: Optional[MCSampler] = None,
+        action_sampler: Optional[MCSampler] = None,
```

Comment:
This has a very different signature than `MCAcquisitionFunction.__init__`. This will make it harder to compare this acquisition function against other acquisition functions or use it within a "glue" framework like Ax that allows for using different acquisition functions with a common interface. It will also make static type checkers like Mypy or Pyre fail.

We have not been requiring type checkers to pass on BoTorch, and we already have a lot of acquisition functions that don't adhere to the common interface defined by `MCAcquisitionFunction` and `AcquisitionFunction`, so this signature isn't doing anything unusual. However, it would be helpful if the signature were as similar as possible to the signature for [`MCAcquisitionFunction.__init__`](https://github.com/pytorch/botorch/blob/646b6c1ac3d5c8c7c4e9af2c835be23478db84e6/botorch/acquisition/monte_carlo.py#L56). Is that possible?

Reply:
The H-Entropy Search acquisition function needs a loss function that is tailored to the problem at hand. However, incorporating this function into the initialization process may potentially disrupt the signature of MCAcquisitionFunction.init. If you have any insights on how to circumvent this issue, please let me know. I would really like to keep the signature of H-Entropy Search consistent with MCAcquisitionFunction as well.